### DISCUSSION TITLE
Global state resets on client side reload of page

### DISCUSSION THREAD

**Discussion by ghost** [vercel/next.js#17615] [Help] [answered]
I am using a context api in the _app.js and passing in a jwt token in all my pages in the application. The problem is that the state persists whenever I am using client side navigation from the application.

But when I do a refresh manually from the browser, the context api prop becomes null. 

How to make sure that the state persists even when I am doing a browser side refresh?

**Comment by mxmtsk** [comment]
You can store the token in a cookie or localstorage and check on mount of the app. if there is a token stored somewhere you hydrate your context with this data.

**Comment by ghost** [comment]
`import "react-multi-carousel/lib/styles.css";
import "../global.css";
import UserContext from "../utils/contextShare";

function MyApp({ Component, pageProps }) {
  const [authUser, setAuthUser] = React.useState(null);
  const [pageLoading, setPageLoading] = React.useState(false);

  React.useEffect(() => {
    const user = localStorage.getItem("user");
    if (user) {
      setAuthUser(user);
    }
  }, []);
  return (
    <>
      <UserContext.Provider value={{ user: authUser }}>
        <Component {...pageProps} />
      </UserContext.Provider>
    </>
  );
}

export default MyApp;
`

This is what my _app.js looks like. What am I doing wrong here?

**Reply by mxmtsk** [reply]
@kamalmohit Are you storing the user in localstorage as well when he logs in?

**Reply by ghost** [reply]
`export const userLogin = async (username, password) => {
  const user = {
    type: "string",
    id: username,
    password: password,
  };

  try {
    const res = await axios.post("/auth/login", user);
    if (res) {
      localStorage.setItem("user", JSON.stringify(res.data.token));
      return res.data.token;
    }
  } catch (error) {
    return false;
  }
};`

this is where I am setting my token in the localstorage

**Comment by lfades** [comment, marked as answer]
@kamalmohit What you're asking for is not possible, there's no "global state" you can re-use after a page refresh, everything in `_app` is only for the current browser session. If you need to persist something for the same browser, use cookies, localStorage, or a DB

**Reply by ghost** [reply]
I am not trying to maintain a global state. I just want my jwt token to be available to all my pages in the Nextjs app, even if I am reloading from the client browser.

**Reply by ghost** [reply]
Please see my above code snippet to understand what I am trying to say. @lfades

**Reply by lfades** [reply]
@kamalmohit If the token is in `localStorage` it will be available for all of your pages, even with a browser reload. Just keep in mind that it will only be available after hydration (i.e use it inside `useEffect`)

**Reply by lfades** [reply]
The context object will always reset (its value is on memory), after every refresh you have to set its value

**Reply by ghost** [reply]
Can you share a code sandbox snippet of what you are saying right here, if you dont mind?

**Reply by ghost** [reply]
Are you suggesting that I dont use a context provider and instead just pass in the value of localstorage directly via prop?

**Reply by ghost** [reply]
Like `<Component user = <token set here> {...pageProps} />`?

**Reply by lfades** [reply]
@kamalmohit What you're doing is okay, setting the context to the value in localStorage is okay, it's something you have to do after every refresh because context saves the value in memory, while localStorage is a browser store.

**Reply by ghost** [reply]
Can you share some code snippets as I am having a hard time grasping this concept. How can I set the context api value on every refresh based on my original code format?

**Reply by mxmtsk** [reply]
Something like this should work:

```js
const UserProvider = ({children}) => {
  const [user, setUser] = React.useState(null);

  React.useEffect(() => {
    // hydrate on mount
    const user = localStorage.getItem("user");
    if (user) {
      setAuthUser(user);
    }
  }, []);

  return (
    <UserContext.Provider value={{ user, setUser }}>
      {children}
   </UserContext.Provider> 
  );
}
```

and in `_app` instead of using `UserContext.Provider` you use the new component `UserProvider`

**Reply by ghost** [reply]
This worked. Thanks!
